Update to GHC 9.2.7

abailly-iohk opened the pull request:
This PR upgrades build infrastructure to use GHC 9.2.7 which seems more stable and more efficient on all platforms. I was able to build the code and run (most of) the tests on Darwin with ARM M2 chip which was not working before. 

---

@ch1bo adopted this work

* Fixed the compilation of `strict-containers` also for the nix tooling.

* Mentioned that this is a **BREAKING** change to the hydra scripts.

---

<!-- Tick off or strike-through / remove if not applicable -->
* [x] CHANGELOG updated
* [x] Documentation updated
* [x] Added and/or updated haddocks
* [x] No new TODOs introduced or explained herafter

abailly committed:
Adapt code to newest GHC version

Seems like GHC got smarter at infering constraints from associated
type families, but stricter on duplicate record fields' ambiguity.

v0d1ch commented:
Not sure if it helps but this is the error I get locally related to `typerep-map` package building:
```
[24/0/89 built] building typerep-map-lib-typerep-map-0.3.0-haddock (configurePhase): --prefix=/nix/store/ik03rsdsglqnk9h7ss0kqc4ywk32rr8y-typerep-map-lib-typerep-map-0.3.0 lib:typerep-map --package-db=clear --package-db=/nix/store/x4f202g0f6f9l36yfkybijr72arp2a6w-typerep-map-lib-typeredirenv: ([/nix/store/s8g46aq00vx3n89awwa7va52jkyiqjcb-direnv-2.32.1/bin/direnv export fish]) is taking a while to execute. Use CTRL-C to give up.
error: builder for '/nix/store/rf9cc0h15dyij8f42a86gf53ksma4qjk-strict-containers-lib-strict-containers-0.1.0.0.drv' failed with exit code 1;
       last 10 log lines:
       >     in the instance declaration for ‘Monad StrictMaybe’.
       >     ‘return’ will eventually be removed in favour of ‘pure’
       >     Either remove definition for ‘return’ (recommended) or define as ‘return = pure’
       >     See also: https://gitlab.haskell.org/ghc/ghc/-/wikis/proposal/monad-of-no-return
       >    |
       > 64 |   return = SJust
       >    |   ^^^^^^^^^^^^^^
       > [2 of 4] Compiling Data.Unit.Strict ( src/Data/Unit/Strict.hs, dist/build/Data/Unit/Strict.o, dist/build/Data/Unit/Strict.dyn_o )
       > [3 of 4] Compiling Data.Sequence.Strict ( src/Data/Sequence/Strict.hs, dist/build/Data/Sequence/Strict.o, dist/build/Data/Sequence/Strict.dyn_o )
       > [4 of 4] Compiling Data.FingerTree.Strict ( src/Data/FingerTree/Strict.hs, dist/build/Data/FingerTree/Strict.o, dist/build/Data/FingerTree/Strict.dyn_o )
       For full logs, run 'nix log /nix/store/rf9cc0h15dyij8f42a86gf53ksma4qjk-strict-containers-lib-strict-containers-0.1.0.0.drv'.
error: 1 dependencies of derivation '/nix/store/i5ccpw8p6swwx9yzymq0zmj38zgw9bwy-ghc-shell-for-packages-env.drv' failed to build

```

uhbif19 commented:
@v0d1ch @abailly-iohk 

Well it seems like `strict-containers` version should be bumped to `0.2`, cuz it claims support of GHC up to 9.4.

Last version of `typerep-map` claim (in cabal file) to support that GHC version as well.

uhbif19 commented:
@v0d1ch @abailly-iohk @ch1bo 

Any updates on this? Seems like simple error with old packages, but I do not know how to debug that in Nix further, cuz it does not give me full cabal output.

ch1bo commented:
> @v0d1ch @abailly-iohk @ch1bo
> 
> Any updates on this? Seems like simple error with old packages, but I do not know how to debug that in Nix further, cuz it does not give me full cabal output.

No update as nobody is working on this right now. To get the full output of a failed nix build you can use the `log-lines` option on the command or put `log-lines = 1000` in your `nix.conf`

ch1bo committed:
Fix compilation of strict-containers

ch1bo committed:
Define default package to be hydra-node

ch1bo commented:
Adopted this work and fixed the nix build. Currently running tests and this should be soon reviewable.

ch1bo committed:
Add changelog entry as this is a breaking change to hydra scripts

Changing the GHC compiler, changes the way the plutus-tx plugin is
executed and hence the resulting plutus-core script that is put on
chain.

ch1bo committed:
Use GHC 9.2.7 in cabal-only CI workflow

v0d1ch reviewed (approved):
Would be nice not to swallow the warnings related to ambiguous fields 🤓

ltouro commented:
should this change dev setup too?

`.github/workflows/actions/setup_dev_ubuntu/action.yaml#7` 
```yaml
name: 'Setup dev environment'
description: 'Setup a dev environment for gh-action Ubuntu'
inputs:
  version_ghc:
    description: 'GHC version needed'
    required: false
    default: 8.10.7
``` 
`.github/workflows/actions/setup_dev_ubuntu/install_packages#45` 
```sh
# Install ghc and cabal
version_ghc=${VERSION_GHC:-8.10.7}
```

ffakenz commented on hydra-cluster/src/Hydra/Cluster/Scenarios.hs:
```diff
@@ -1,5 +1,6 @@
 {-# LANGUAGE DuplicateRecordFields #-}
 {-# LANGUAGE TypeApplications #-}
+{-# OPTIONS_GHC -Wno-ambiguous-fields #-}
```
why is this needed?

ffakenz commented on hydra-node/hydra-node.cabal:
```diff
@@ -345,7 +341,6 @@ test-suite tests
     , io-sim
     , lens
     , lens-aeson
-    , network
```
interesting.. so network is also included by default 😮

ffakenz commented on hydra-tui/test/Hydra/TUISpec.hs:
```diff
@@ -1,4 +1,5 @@
 {-# LANGUAGE DuplicateRecordFields #-}
+{-# OPTIONS_GHC -Wno-ambiguous-fields #-}
```
why is this needed?

v0d1ch reviewed (approved):
Works for me! :tada:
